src/store/kv-block-store.ts

```diff
@@ -0,0 +1,205 @@
+        message: error.message,
+        stack: error.stack,
+      });
+    }
+    return undefined;
+  }
+
+  async getByHeight(height: number): Promise<PartialJsonBlock | undefined> {
+    try {
+      if (await this.hasHeight(height)) {
+        const blockHashBuffer = await this.kvBufferStore.get(
+          this.blockHeight(height),
+        );
+        if (blockHashBuffer === undefined) {
+          throw new Error('Missing block hash in key/value store for height.');
+        }
+        const blockHash = toB64Url(blockHashBuffer);
+        return this.getByHash(blockHash);
+      }
+    } catch (error: any) {
+      this.log.error('Failed to get block by height', {
+        height,
+        message: error.message,
+        stack: error.stack,
+      });
+    }
+    return undefined;
+  }
+
+  async delByHash(hash: string): Promise<void> {
+    try {
+      if (await this.hasHash(hash)) {
+        const blockData = await this.getByHash(hash);
+        // remove the hash to block data reference
+        await this.kvBufferStore.del(this.blockHash(hash));
+
+        // remove the block height to hash reference
+        if (
+          blockData !== undefined &&
+          (await this.hasHeight(blockData.height))
+        ) {
+          // TODO: should we check the hash stored in the table matches the blockData, or delete regardless
+          await this.kvBufferStore.del(this.blockHeight(blockData.height));
+        }
+      }
+    } catch (error: any) {
+      this.log.error('Failed to delete block by height', {
+        hash,
+        message: error.message,
+        stack: error.stack,
+      });
+    }
+  }
+
+  async delByHeight(height: number): Promise<void> {
+    try {
+      if (await this.hasHeight(height)) {
+        const blockHashBuffer = await this.kvBufferStore.get(
+          this.blockHeight(height),
+        );
+        // remove height to block hash reference
+        await this.kvBufferStore.del(this.blockHeight(height));
+
+        // remove the block hash to block data reference
+        if (blockHashBuffer !== undefined) {
+          const blockHash = toB64Url(blockHashBuffer);
+          if (await this.hasHash(blockHash)) {
+            await this.kvBufferStore.del(this.blockHash(blockHash));
+          }
+        }
+      }
+    } catch (error: any) {
+      this.log.error('Failed to delete block by height', {
+        height,
+        message: error.message,
+        stack: error.stack,
+      });
+    }
+  }
+
+  async set(block: PartialJsonBlock, height?: number): Promise<void> {
+    const hash = block.indep_hash;
+    try {
+      if (!(await this.hasHash(hash))) {
+        const blockData = jsonBlockToMsgpack(block);
+        await this.kvBufferStore.set(this.blockHash(hash), blockData);
+
+        // store the height against the hash value to avoid duplication of block data in the KV store
+        if (height !== undefined && !(await this.hasHeight(height))) {
+          const encodedHash = fromB64Url(hash);
```

Comment:
Nit: should this be named `hashBuffer` buffer rather than `decodedHash`?

Reply:
[66f4cf7](https://github.com/ar-io/ar-io-node/pull/60/commits/66f4cf7800807ed7f587e875fbde7653a103bb18) 😄